File: construct_erpnext/unit_costing/report/unit_cost_allocation_report/unit_cost_allocation_report.py
import frappe
from frappe import _


def execute(filters=None):
	filters = filters or {}
	columns = [
		{"label": _("Allocation"), "fieldname": "name", "fieldtype": "Link", "options": "Unit Cost Allocation", "width": 140},
		{"label": _("Project"), "fieldname": "real_estate_project", "fieldtype": "Link", "options": "Real Estate Project", "width": 180},
		{"label": _("Basis"), "fieldname": "allocation_basis", "fieldtype": "Data", "width": 120},
		{"label": _("Cost Source"), "fieldname": "cost_source", "fieldtype": "Data", "width": 130},
		{"label": _("Source Amount"), "fieldname": "source_amount", "fieldtype": "Currency", "width": 130},
		{"label": _("Total Allocated"), "fieldname": "total_allocated_amount", "fieldtype": "Currency", "width": 140},
		{"label": _("Unallocated"), "fieldname": "unallocated_amount", "fieldtype": "Currency", "width": 120},
		{"label": _("Status"), "fieldname": "allocation_status", "fieldtype": "Data", "width": 110},
		{"label": _("Date"), "fieldname": "allocation_date", "fieldtype": "Date", "width": 110},
	]
	conditions = ["1=1"]
	values = {}
	for field in ("real_estate_project", "allocation_status", "allocation_basis", "cost_source"):
		if filters.get(field):
			conditions.append(f"{field} = %({field})s")
			values[field] = filters[field]
	data = frappe.db.sql(
		f"""
		SELECT name, real_estate_project, allocation_basis, cost_source, source_amount,
			total_allocated_amount, unallocated_amount, allocation_status, allocation_date
		FROM `tabUnit Cost Allocation`
		WHERE {" AND ".join(conditions)}
		ORDER BY allocation_date DESC, creation DESC
		""",
		values,
		as_dict=True,
	)
	return columns, data

File: construct_erpnext/unit_costing/report/unit_cost_allocation_report/unit_cost_allocation_report.js
frappe.query_reports["Unit Cost Allocation Report"] = {
	filters: [
		{fieldname: "real_estate_project", label: __("Real Estate Project"), fieldtype: "Link", options: "Real Estate Project"},
		{fieldname: "allocation_status", label: __("Allocation Status"), fieldtype: "Select", options: "\nDraft\nCalculated\nReviewed\nApplied\nCancelled"},
		{fieldname: "allocation_basis", label: __("Allocation Basis"), fieldtype: "Select", options: "\nBy Area\nEqual Share\nManual Percentage\nManual Amount"},
		{fieldname: "cost_source", label: __("Cost Source"), fieldtype: "Select", options: "\nBOQ Total\nCommitted Cost\nInvoiced Cost\nConsumed Cost\nCertified Cost\nFinancial Snapshot\nManual Amount"}
	]
};
